crates/bevy_ecs/src/world/entity_access/entity_mut.rs

```diff
@@ -186,11 +186,42 @@ impl<'w> EntityMut<'w> {
     /// the `QueryData` does not provide aliasing mutable references to the same component.
     pub unsafe fn get_components_mut_unchecked<Q: ReleaseStateQueryData>(
         &mut self,
-    ) -> Option<Q::Item<'_, 'static>> {
-        // SAFETY: Caller the `QueryData` does not provide aliasing mutable references to the same component
+    ) -> Result<Q::Item<'_, 'static>, QueryAccessError> {
+        // SAFETY: Caller ensures the `QueryData` does not provide aliasing mutable references to the same component
         unsafe { self.reborrow().into_components_mut_unchecked::<Q>() }
     }
 
+    /// Returns components for the current entity that match the query `Q`.
+    /// In the case of conflicting [`QueryData`](crate::query::QueryData), unregistered components, or missing components,
+    /// this will return a [`QueryAccessError`]
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use bevy_ecs::prelude::*;
+    /// #
+    /// #[derive(Component)]
+    /// struct X(usize);
+    /// #[derive(Component)]
+    /// struct Y(usize);
+    ///
+    /// # let mut world = World::default();
+    /// let mut entity = world.spawn((X(0), Y(0))).into_mutable();
+    /// // Get mutable access to two components at once
+    /// // SAFETY: X and Y are different components
+    /// let (mut x, mut y) = entity.get_components_mut::<(&mut X, &mut Y)>().unwrap();
+    /// ```
+    ///
+    /// Note that this does a O(n^2) check that the [`QueryData`](crate::query::QueryData) does not conflict. If performance is a
+    /// consideration you should use [`Self::get_components_mut_unchecked`] instead.
+    pub fn get_components_mut<Q: ReleaseStateQueryData>(
```

Comment:
We may also want `into_components_mut` versions of these that consume `self` and return the full `'w` lifetime.  

You can share the implementations by implementing `EntityMut::get_components_mut` as `self.reborrow().into_components_mut::<Q>()` and `EntityWorldMut::get_components_mut` as `self.as_mutable().into_components_mut::<Q>()`.  That also makes it harder to accidentally extend the lifetimes, since you're delegating that to the existing `reborrow()` and `as_mutable()` methods.  (That's what I did in #20273.)